### PR TITLE
Add support for <SID>

### PR DESCRIPTION
**Pull request by wsdjeg**
test with:
```vim
fu! s:test()
endf

let test = <SID>test()
```

cursor is on `<SID>test()`

### CODE REVIEW

File: autoload/lookup.vim [hunk 1]
```diff
@@ -44,7 +44,7 @@ endfunction
 
 " s:find_local_func_def() {{{1
 function! s:find_local_func_def(funcname) abort
-  if search('\c\v<fu%[nction]!?\s+%(s:|\<sid\>)\zs\V'.a:funcname.'\>', 'bsw') != 0
+  if search('\c\v<fu%[nction]!?\s+%(s:|\<sid\>|\<SID\>)\zs\V'.a:funcname.'\>', 'bsw') != 0
```
Comment (mhinz) [inline autoload/lookup.vim:47]: This shouldn't be needed because of the leading `\c`.

File: autoload/lookup.vim [hunk 2]
```diff
@@ -14,7 +14,7 @@ function! lookup#lookup() abort
   let &iskeyword = isk
   let is_func = name =~ '($' ? 1 : 0
   let could_be_funcref = name =~ '[''"]$' ? 1 : 0
-  let name = matchstr(name, '\v^%(s:|\<sid\>)?\zs.{-}\ze[\("'']?$')
+  let name = matchstr(name, '\v^%(s:|\<sid\>|\<SID\>)?\zs.{-}\ze[\("'']?$')
```
Comment (mhinz) [inline autoload/lookup.vim:17]: Can you please put a `\c` in front of the `\v` instead?

File: autoload/lookup.vim (hunk 1 above)
Comment (wsdjeg) [reply autoload/lookup.vim:47]: I think maybe we should not use `\c`, here only `sid` is ignorecase, `fu` `function` and `s:` is not ignorecase.

Comment (mhinz) [reply autoload/lookup.vim:47]: But `S:` etc. would be invalid syntax anyway. The problem is that this has to work as well:

```vim
function! <sId>f()
endfunction

call <SiD>f()
```

**Comment by mhinz** [comment]
BTW, your example works for me on master. :o

**Comment by wsdjeg** [comment]
It works well on master? But I can not get it work. i will test more tonight.

**Comment by mhinz** [comment]
It should. :] Using your example:

At [this point](https://github.com/mhinz/vim-lookup/blob/54de6b02f94a5b2b5d00f8c30773d7b03cd57bf9/autoload/lookup.vim#L20), `is_auto` should be 0 and `is_func` should be 1. Therefore it will run `s:find_local_func_def('test')`.

The first line in that function is a `search()` that should find the function definition already and then returns: https://github.com/mhinz/vim-lookup/blob/54de6b02f94a5b2b5d00f8c30773d7b03cd57bf9/autoload/lookup.vim#L48

**Comment by wsdjeg** [comment]
@mhinz I have test in master, but it does not work, here is an example:
```vim
  nnoremap <silent> <Plug>CommentToLine :call <sid>comment_to_line(0)<Cr>
  nnoremap <silent> <Plug>CommentToLineInvert :call <SID>comment_to_line(1)<Cr>
function! s:comment_to_line(invert) abort
  let input = input('line number :')
  if empty(input)
    return
  endif
  let line = str2nr(input)
  let ex = line - line('.')
  if ex > 0
    exe 'normal! V'. ex .'j'
  elseif ex == 0
  else
    exe 'normal! V'. abs(ex) .'k'
  endif
  if a:invert
    call feedkeys("\<Plug>NERDCommenterInvert")
  else
    call feedkeys("\<Plug>NERDCommenterComment")
  endif
endfunction
```

move cursor on  `comment_to_line` in line 2, and  it does not work, but on line 1 it works well.

after patching this PR, it works well.

**Comment by mhinz** [comment]
Hmm, I don't get it. Lines 1 and 2 work for me. The cursor jumps to line 3.

Can you try again by using `:call lookup#lookup()` on line 2 again after applying this diff?

```vim
diff --git i/autoload/lookup.vim w/autoload/lookup.vim
index a70a3b8..909807c 100644
--- i/autoload/lookup.vim
+++ w/autoload/lookup.vim
@@ -13,0 +14 @@ function! lookup#lookup() abort
+  echomsg name
@@ -17,0 +19 @@ function! lookup#lookup() abort
+  echomsg name
```

It should output:

```
<SID>comment_to_line(
comment_to_line
```

**Comment by wsdjeg** [comment]
no, it is : 
```
<SID>comment_to_line(                                                                                                                                  
<SID>comment_to_line
```

**Comment by mhinz** [comment]
Hmmmmm. That's weird.

How can the output be different for us?

```
:echo matchstr('<SID>comment_to_line(', '\v^%(s:|\<sid\>)?\zs.{-}\ze[\("'']?$')
```

**Comment by mhinz** [comment]
Oooh. You have `set noignorecase`?

**Comment by mhinz** [comment]
Yup, that's it. Together with the reason mentioned earlier (matching `<SiD>` etc.), please edit the commit to just do this:

```diff
diff --git i/autoload/lookup.vim w/autoload/lookup.vim
index a70a3b8..39300e4 100644
--- i/autoload/lookup.vim
+++ w/autoload/lookup.vim
@@ -14,7 +14,7 @@ function! lookup#lookup() abort
   let &iskeyword = isk
   let is_func = name =~ '($' ? 1 : 0
   let could_be_funcref = name =~ '[''"]$' ? 1 : 0
-  let name = matchstr(name, '\v^%(s:|\<sid\>)?\zs.{-}\ze[\("'']?$')
+  let name = matchstr(name, '\c\v^%(s:|\<sid\>)?\zs.{-}\ze[\("'']?$')
   let is_auto = name =~ '#' ? 1 : 0
   let position = s:getcurpos()
   if !dispatch[is_auto][is_func](name) && !is_func && could_be_funcref
```

**Comment by wsdjeg** [comment]
@mhinz  this patch works well. 👍

**Comment by mhinz** [comment]
Oh, should I commit it?

**Comment by mhinz** [comment]
I altered your commit from this PR to add the change. ;-)

**Comment by wsdjeg** [comment]
Wow. Thanks. :)

**Pull request closed without merging**